Add MARTI dice server to source control (#8)

ssoloff opened the pull request:
This PR addresses issue #8, and adds the MARTI dice server to source control [without any sensitive information embedded in the source code](https://12factor.net/config).

#### Functional changes

None.

#### Refactoring changes

* The hard-coded database connection information was removed from `dice.class.php` and into the environment variables `MARTI_DB_HOST`, `MARTI_DB_USERNAME`, `MARTI_DB_PASSWORD`, and `MARTI_DB_NAME`.  The `dice` class was updated to retrieve this information from the environment when establishing a database connection (see the `dice::connectDatabase()` method and commit [7b1b76e](https://github.com/ssoloff/triplea-game-tripleawarclub.org/commit/7b1b76ebb793adb298d4f09cdd555d335b58e7ca#diff-c44244a14c62f0c498a48a315c7d7ac6)).
* The hard-coded base URI for all links was removed from `dice.class.php` and calculated from information in the active request (see commit c7df97c).  This ensures links in emails sent by MARTI always reflect how the user accessed the service (e.g. a user can use either `http[s]://dice.tripleawarclub.org/MARTI.php` or `http[s]://www.tripleawarclub.org/dice/MARTI.php` and the email links will be based on the appropriate base URI).
* The encryption key is created if it does not already exist (see commit 94a36a2).  This is primarily for use cases where a developer spins up a Docker container for testing a pristine MARTI install (see below).

#### Refactoring issues for review

* Extracting the `MARTI_DB_*` environment variables will require changes to the production server if this code is deployed to `tripleawarclub.org`.  Basically, one `fastcgi_param` entry needs to be added per environment variable in an appropriate Nginx configuration file (see commit [7b1b76e](https://github.com/ssoloff/triplea-game-tripleawarclub.org/commit/7b1b76ebb793adb298d4f09cdd555d335b58e7ca#diff-1f27f05ab6603611d2ca89a3535efb26)).  Unlike the Dockerfile example at that link, these values can be placed in a separate `*_params` file and included in the top-level Nginx configuration.  Then the separate `*_params` file can be locked down as needed.  **Please advise if there is a more preferable approach to extracting the database connection information other than environment variables.**
* I used the database schema as it exists in production.  We had originally [discussed](https://github.com/triplea-game/tripleawarclub.org/issues/8#issuecomment-298183446) switching to InnoDB tables and the Unicode character set.  Upon reflection, I decided to leave it as in production because the test image should be as close to production as possible in order to reproduce any issues.  When the production database is updated to use this more-desirable configuration, we can update the Docker image, as well.

#### Other changes

* I created a Docker image for testing MARTI.  This is a simple, no-nonsense image.  **It should not be used for production** as it runs all services in a single container and the parent image (`stenote/docker-lemp`) hard-codes database credentials in the image.  I verified that I could use it to test MARTI registration, dice rolling, and roll validation.  See `build/docker/README.md` for usage information.

#### Other issues for review

* I placed all the Docker configuration stuff in the top-level `build/docker` folder.  This will be an issue if Gradle is ever used in this repo.  Is there a folder naming convention the team uses for build-related sources, and should the `build` folder be renamed to something else?  The closest thing I saw in the `triplea-game` repo is the `.travis` folder.  I used `build/docker` because I originally also had a `build/mysql` folder, and I didn't want to pollute the root with multiple new folders.
* Not really part of this change, but I noticed it while working this PR...  MARTI has code to periodically (every month) regenerate its encryption key.  However, this code does not appear to ever run.  In fact, the timestamp on the "current" encryption key file used in production indicates it hasn't changed since April 2008.  Not sure if this is something the team wants to follow up on.

@RoiEXLab @ron-murhammer @DanVanAtta

RoiEXLab commented:
@ssoloff I found out that bung created the tripleawarclub server originally... I asked him about the email config... Still waiting for an answer

ssoloff commented on build/docker/files/root/entrypoint.sh:
```diff
@@ -0,0 +1,55 @@
+chown -R mysql:mysql /var/lib/mysql
+
+if [ ! -d "/var/lib/mysql/$MARTI_DB_NAME" ]; then
+  service mysql start
```
The parent image of my Docker container does not auto start any services controlled by `supervisord` (of which MySQL is one).  So it is guaranteed that MySQL is not running at this point (`supervisord` is not started until the last line in the `entrypoint.sh` script).

ssoloff commented on build/docker/files/root/docker-entrypoint-initdb.d/marti-init.sql:
```diff
@@ -0,0 +1,22 @@
+  `time_stamp` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+  `IP` int(10) unsigned NOT NULL,
+  UNIQUE KEY `email` (`email`)
+) ENGINE=MyISAM DEFAULT CHARSET=latin1;
```
Ah, ok, I was under the impression InnoDB was a suggestion for future use. :smiley:  Since it is already that way in production, I'll update the script to use it.

RoiEXLab replied:
I thought this was the case as well, but you can change the engine with a single command... Something along this: `ALTER TABLE tablename ENGINE=InnoDB;

ssoloff commented:
@RoiEXLab 

> The repo is currently out-of-sync with the warclub server, it has a newer xoops version.

Is MARTI dependent on xoops?  My Docker image only contains the PHP code under `public_html/dice` and MARTI seems to run fine with the rest of the repo's PHP code missing.  Since this PR only focused on MARTI, I didn't really consider code outside the `dice` folder.

> when we have the chance like now, we might just change things!

In general, I agree.  You don't know how many times I caught myself changing minor things (like indentation) while working this PR only to back them out. :smiley:  I had to keep reminding myself that the focus of this PR was to get the MARTI code into source control as it exists today.  I didn't want to introduce a bunch of unrelated changes that would overwhelm the review (well, I guess that's not as much of a problem here because all the files are new and not changes to existing files).

So, while I'm more than happy to make all the formatting changes and improvements you suggested, would it not be better to do them in separate PRs?  I think the most important thing right now is ensuring we can deploy the modified MARTI code (i.e. with the configuration extracted to the environment) from the repo to production.  Once we know we can do that, then I'll feel more confident about making additional changes.

Plus, I would feel more comfortable adding some tests before making the functional changes.  The last time I did any significant PHP work was about eight years ago, and I don't trust myself not to break something. :smiley:

RoiEXLab commented on tripleawarclub.org/public_html/dice/MARTI.php:
```diff
@@ -0,0 +1,68 @@
		// check if all emails are registered
+		try {
+			$dice->checkIfMailsAreRegistered($output['emails']); 
+		} catch(exception $e) {
```
Exception should be spelled with an uppercase letter...
it seems not all PHP versions treat class names case-insensitive

RoiEXLab commented on tripleawarclub.org/public_html/dice/dice.class.php:
```diff
@@ -0,0 +1,287 @@
+	var $db = null;
+	var $enc = [];
+
+	// constructor
```
Obvious comments should be removed...

RoiEXLab commented on tripleawarclub.org/public_html/dice/dice.class.php:
```diff
@@ -0,0 +1,287 @@
+		$this->connectDatabase();
+    }
+
+	// destructor
```
Same here

RoiEXLab commented on tripleawarclub.org/public_html/dice/dice.class.php:
```diff
@@ -0,0 +1,287 @@
+		$sql = "SELECT * FROM stats";
+		$result = $this->dbconn->query($sql) or exit("fatal error: data connection lost @getStats!");
+		$stats = mysqli_fetch_array($result);
+		return $stats;
```
This line can also be merged with the preceding line

RoiEXLab commented on tripleawarclub.org/public_html/dice/MARTI.php:
```diff
@@ -0,0 +1,68 @@
+	<body>
+		<?php
+		error_reporting(E_ALL);
+		ini_set('display_errors', '1');
```
Duplicate code, this is already execute before the doctype is sent...
I think removing the part before the doctype is probably better!

RoiEXLab commented on tripleawarclub.org/public_html/dice/MARTI.php:
```diff
@@ -0,0 +1,68 @@
+		$emails1 = explode(" ", $email1);
+		$emails2 = explode(" ", $email2);
+
+		$output['emails'] = array_merge($emails1, $emails2);
```
This line should be merged with 31 and 32

RoiEXLab commented:
> Is MARTI dependent on xoops? My Docker image only contains the PHP code under public_html/dice and MARTI seems to run fine with the rest of the repo's PHP code missing. Since this PR only focused on MARTI, I didn't really consider code outside the dice folder.

No it's not, but it's located on the server just like it is in the repo. If we wanted to pull the changes via git we may would end up changing the local non-marti files...

> So, while I'm more than happy to make all the formatting changes and improvements you suggested, would it not be better to do them in separate PRs? I think the most important thing right now is ensuring we can deploy the modified MARTI code (i.e. with the configuration extracted to the environment) from the repo to production. Once we know we can do that, then I'll feel more confident about making additional changes.

The problem I see here is that just like on the other repositorys I don't have write access, e.g. I can't merge PRs. Unless @DanVanAtta or @ron-murhammer merge those changes soon, we may end up waiting a long time until a second PR is merged...

> Plus, I would feel more comfortable adding some tests before making the functional changes. The last time I did any significant PHP work was about eight years ago, and I don't trust myself not to break something. 

Legit concern ^^ PHP is a bad mixture of C with Java-like terms with some major design flaws. (Also why are there almost every time 2 ways of doing something?)

RoiEXLab commented:
@ssoloff I will keep adding comments for future reference...
Just ignore them for now if you want them to be implemented in a different PR

RoiEXLab commented on tripleawarclub.org/public_html/dice/dice.class.php:
```diff
@@ -0,0 +1,287 @@
+		$this->connectDatabase();
+
+		$sql = "UPDATE stats SET requests=requests+1, dice_rolled=dice_rolled+$numdice";
+		$result = $this->dbconn->query($sql) or exit("fatal error: data connection lost @updateStats!");
```
Note: This functions seems to no longer be used

RoiEXLab commented on tripleawarclub.org/public_html/dice/dice.class.php:
```diff
@@ -0,0 +1,287 @@
+    	$this->db = mysqli_select_db($this->dbconn, $database);
+    }
+
+	function updateStats($numdice) {
```
Note: This function seems to no longer be used

RoiEXLab commented on tripleawarclub.org/public_html/dice/dice.class.php:
```diff
@@ -0,0 +1,287 @@
+
+	}
+
+	function getStats() {
```
Note: This function seems to no longer be used

RoiEXLab commented on tripleawarclub.org/public_html/dice/dice.class.php:
```diff
@@ -0,0 +1,287 @@
+		$result = $this->dbconn->query($sql) or exit("fatal error: data connection lost @checkIfMailsAreRegistered!");
+		$registered_mails = mysqli_fetch_array($result);
+		$num_emails = $result->num_rows;
+		if($num_emails == count($emails)){
```
Line can be merged with preceding line

RoiEXLab commented on tripleawarclub.org/public_html/dice/dice.class.php:
```diff
@@ -0,0 +1,287 @@
+	 * Checks if all receiving email adresses are registered
+	 * @returns bool
+	 */
+	function checkIfMailsAreRegistered(array $emails) {
```
Future Note: I'd argue that `checkIfMailsAreRegistered` does the same as `checkIfMailIsRegistered`...

RoiEXLab commented on tripleawarclub.org/public_html/dice/dice.class.php:
```diff
@@ -0,0 +1,287 @@
+	function runStatement($sth){
+		$this->connectDatabase();
+		$result = $sth->execute();
+		return $result;
```
This line should be merged with preceding line

RoiEXLab commented on build/docker/files/root/docker-entrypoint-initdb.d/marti-init.sql:
```diff
@@ -3,16 +3,6 @@ CREATE TABLE `dice_emails` (
-  PRIMARY KEY (`request_ID`),
-  KEY `UUID` (`UUID`)
-) AUTO_INCREMENT=57 DEFAULT CHARSET=latin1;
-
```
Yeah, noticed that as well...
There is an empty database on the warclub server ^^

ssoloff commented:
@RoiEXLab 

I updated the engine to InnoDB.  I also removed the schema for the `dice_table` table.  It does not appear to be used in the current codebase.

> If we wanted to pull the changes via git we may would end up changing the local non-marti files...

Agreed for a full deployment from Git.  I was assuming that since _nothing_ is currently deployed from Git, the first step would be to deploy just the `dice` folder when that code is updated.  A Git clone would probably be overkill anyway.  Simply downloading the latest tarball and copying the `dice` folder should be sufficient.

> I will keep adding comments for future reference...

Yes, please do while they're fresh in your mind.

RoiEXLab commented on tripleawarclub.org/public_html/dice/dice.class.php:
```diff
@@ -0,0 +1,287 @@
+		}
+	}
+
+    function keygen() {
```
I believe there's a more efficient way of generating a random password of length 24 matching `[A-Za-z0-9]{24}`

RoiEXLab commented on tripleawarclub.org/public_html/dice/dice.class.php:
```diff
@@ -0,0 +1,287 @@
+			$encrypt_key = $this->getEncryptionKey($date);
+		} else {
+			$encrypt_key = $this->getEncryptionKey();
+		}
```
This if else statement can be shortened into a short assignment I forgot what it's actually called: `$encrypt_key = $this->getEncryptionKey(($date !== $this->getDate()) ? $date : null);`

RoiEXLab commented on tripleawarclub.org/public_html/dice/dice.class.php:
```diff
@@ -0,0 +1,287 @@
+		while ($i <= $numdice-1) {
+			$dice[$i] = mt_rand(1,$numsides);
+			$i++;
+		}
```
A for loop would be more appropiate here...
Also `$i <= $numdice-1` ? Are you for real? Why?!?!

RoiEXLab commented on tripleawarclub.org/public_html/dice/register.php:
```diff
@@ -0,0 +1,85 @@
+
+			// check if the email already exists
+			if($dice->checkIfMailIsRegistered($email))
+				exit("This email is already registered");
```
Curly brackets

RoiEXLab commented on tripleawarclub.org/public_html/dice/validate.php:
```diff
@@ -0,0 +1,56 @@
+					$sth->execute() or trigger_error($mysqli->error);
+					$sth->bind_result($r_email,$r_validation);
+					while($sth->fetch()){
+							$rows[] = array( $r_email, $r_validation );
```
I think using `[...]` is better than the `array(...)` function

RoiEXLab commented on tripleawarclub.org/public_html/dice/validate.php:
```diff
@@ -0,0 +1,56 @@
+							$rows[] = array( $r_email, $r_validation );
+					}
+			} else {
+					echo "A DB error has occured, please contact an admin. (1-"; var_dump( $dice->dbconn->errno ); echo ")";exit;
```
Formatting

ssoloff commented:
@ron-murhammer @DanVanAtta Bumping...

ssoloff commented:
@ron-murhammer @DanVanAtta Bumping...

ron-murhammer commented:
@ssoloff This looks good. My only question is around whether this is the right repo to put it. The tripleawarclub.org website code is effectively dead now with the new forum. On the other hand, the dice server is very much relevant and I won't consider it really part of tripleawarclub.org. I'd consider it more aligned to the lobby since they are both actively used components hosted on our server. Thoughts? Maybe this belongs in a separate repo all together?

ssoloff commented:
@ron-murhammer @RoiEXLab would know better than me, but I believe MARTI is still running on this host.  The current `dice_servers/tripleawarclub.properties` in triplea-game/triplea is pointing to `dice.tripleawarclub.org`.

@RoiEXLab Can you please confirm where MARTI is hosted?  Also, as Ron pointed out, is there a more appropriate repo for this code?

RoiEXLab commented:
@ssoloff It is running on the same server, using the same nginx process, that is correct, but I agree with @ron-murhammer that this code should be moved to a new Repo!
There you go: https://github.com/triplea-game/dice-server (I just created this repo)

ssoloff commented:
@RoiEXLab Ok, I'll open up a PR against that repo instead.  Thanks!

Since you are the one who will most likely deploy this code into production, how do you want me to structure the project?  Currently, MARTI would be under `/public_html/dice` based on the `tripleawarclub.org` layout.  Do you want me to keep that path or do you prefer something else?

RoiEXLab commented:
@ssoloff Good qustion... I believe using just dice as a folder would be fine.

ssoloff commented:
Closing in favor of triplea-game/dice-server#1.
